cpp/src/parquet/geometry_util_internal.h

```diff
@@ -0,0 +1,395 @@
+      Coord coord;
+      for (uint32_t i = 0; i < n_coords; i++) {
+        coord = ::arrow::util::SafeLoadAs<Coord>(data_);
+        for (uint32_t j = 0; j < coord.size(); j++) {
+          coord[j] = ::arrow::bit_util::ByteSwap(coord[j]);
+        }
+
+        func(coord);
+        data_ += sizeof(Coord);
+        size_ -= sizeof(Coord);
+      }
+    } else {
+      for (uint32_t i = 0; i < n_coords; i++) {
+        func(::arrow::util::SafeLoadAs<Coord>(data_));
+        data_ += sizeof(Coord);
+        size_ -= sizeof(Coord);
+      }
+    }
+
+    return ::arrow::Status::OK();
+  }
+
+  size_t size() { return size_; }
+
+ private:
+  const uint8_t* data_;
+  size_t size_;
+};
+
+struct BoundingBox {
+  BoundingBox(const std::array<double, 4>& mins, const std::array<double, 4>& maxes) {
+    std::memcpy(min, mins.data(), sizeof(min));
+    std::memcpy(max, maxes.data(), sizeof(max));
+  }
+  BoundingBox() : min{kInf, kInf, kInf, kInf}, max{-kInf, -kInf, -kInf, -kInf} {}
+
+  BoundingBox(const BoundingBox& other) = default;
+  BoundingBox& operator=(const BoundingBox&) = default;
+
+  void UpdateXY(std::array<double, 2> coord) { UpdateInternal(coord); }
+
+  void UpdateXYZ(std::array<double, 3> coord) { UpdateInternal(coord); }
+
+  void UpdateXYM(std::array<double, 3> coord) {
+    min[0] = std::min(min[0], coord[0]);
+    min[1] = std::min(min[1], coord[1]);
+    min[3] = std::min(min[3], coord[2]);
+    max[0] = std::max(max[0], coord[0]);
+    max[1] = std::max(max[1], coord[1]);
+    max[3] = std::max(max[3], coord[2]);
+  }
+
+  void UpdateXYZM(std::array<double, 4> coord) { UpdateInternal(coord); }
+
+  void Reset() {
+    for (int i = 0; i < 4; i++) {
+      min[i] = kInf;
+      max[i] = -kInf;
+    }
+  }
+
+  void Merge(const BoundingBox& other) {
+    for (int i = 0; i < 4; i++) {
+      min[i] = std::min(min[i], other.min[i]);
+      max[i] = std::max(max[i], other.max[i]);
+    }
+  }
+
+  std::string ToString() const {
+    std::stringstream ss;
+    ss << "BoundingBox [" << min[0] << " => " << max[0];
+    for (int i = 1; i < 4; i++) {
+      ss << ", " << min[i] << " => " << max[i];
+    }
+
+    ss << "]";
+
+    return ss.str();
+  }
+
+  double min[4];
+  double max[4];
+
+ private:
+  // This works for XY, XYZ, and XYZM
+  template <typename Coord>
+  void UpdateInternal(Coord coord) {
+    static_assert(coord.size() <= 4);
+
+    for (size_t i = 0; i < coord.size(); i++) {
+      min[i] = std::min(min[i], coord[i]);
+      max[i] = std::max(max[i], coord[i]);
+    }
+  }
+};
+
+inline bool operator==(const BoundingBox& lhs, const BoundingBox& rhs) {
+  return std::memcmp(lhs.min, rhs.min, sizeof(lhs.min)) == 0 &&
+         std::memcmp(lhs.max, rhs.max, sizeof(lhs.max)) == 0;
+}
+
+class WKBGeometryBounder {
+ public:
+  WKBGeometryBounder() = default;
+  WKBGeometryBounder(const WKBGeometryBounder&) = default;
+
+  ::arrow::Status ReadGeometry(WKBBuffer* src, bool record_wkb_type = true) {
```

Comment:
from an API perspective is it intended to let callers change record_wkb_type?  If not consider make ReadGeometry without this parameter then move this implementation to a private helper?